homeassistant/components/openhome/media_player.py

```diff
@@ -19,34 +20,81 @@
     SUPPORT_VOLUME_SET,
     SUPPORT_VOLUME_STEP,
 )
-from homeassistant.const import STATE_IDLE, STATE_OFF, STATE_PAUSED, STATE_PLAYING
+from homeassistant.const import (
+    ATTR_ENTITY_ID,
+    STATE_IDLE,
+    STATE_OFF,
+    STATE_PAUSED,
+    STATE_PLAYING,
+)
+import homeassistant.helpers.config_validation as cv
+
+from .const import ATTR_PIN_INDEX, DATA_OPENHOME, DOMAIN, SERVICE_INVOKE_PIN
+
+OPENHOME_PIN_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_id,
+        vol.Required(ATTR_PIN_INDEX): cv.positive_int,
+    }
+)
 
 SUPPORT_OPENHOME = SUPPORT_SELECT_SOURCE | SUPPORT_TURN_OFF | SUPPORT_TURN_ON
 
 _LOGGER = logging.getLogger(__name__)
 
-DEVICES = []
-
 
 def setup_platform(hass, config, add_entities, discovery_info=None):
     """Set up the Openhome platform."""
 
     if not discovery_info:
         return True
 
+    if DATA_OPENHOME not in hass.data:
+        hass.data[DATA_OPENHOME] = []
+
     name = discovery_info.get("name")
     description = discovery_info.get("ssdp_description")
+
     _LOGGER.info("Openhome device found: %s", name)
     device = Device(description)
 
     # if device has already been discovered
-    if device.Uuid() in [x.unique_id for x in DEVICES]:
+    if device.Uuid() in [x.unique_id for x in hass.data[DATA_OPENHOME]]:
         return True
 
-    device = OpenhomeDevice(hass, device)
+    hass_device = OpenhomeDevice(hass, device)
+
+    add_entities([hass_device], True)
+    hass.data[DATA_OPENHOME].append(hass_device)
+
+    def service_handle(call):
+        """Handle the service call."""
 
-    add_entities([device], True)
-    DEVICES.append(device)
+        entity_id = call.data.get(ATTR_ENTITY_ID)
+
+        device = next(
+            [
+                device
+                for device in hass.data[DATA_OPENHOME]
+                if device.entity_id == entity_id
+            ].__iter__(),
+            None,
+        )
+
+        if not device:
+            return
+
+        if call.service == SERVICE_INVOKE_PIN:
+            index = call.data.get(ATTR_PIN_INDEX)
+
+            _LOGGER.info("Openhome invoking pin %s on %s", index, entity_id)
+
+            if index:
+                device.invoke_pin(index)
+
+    hass.services.register(
```

Comment:
Please use the entity service helper instead:
https://developers.home-assistant.io/docs/dev_101_services#entity-services

This requires to refactor the platform to use async context.